---
title: "Reconstruction template 2"
output: html_document
---

---
title: "Genome Reconstruction"
output: html_document
---


## Setting up

Create these directories if needed
```{bash}
# Create directories for the WGS input and BLAST database
mkdir data
mkdir db
mkdir db/fna
mkdir db/blast_db

# Create directories for outputs for each module
mkdir output
mkdir output/module_1
mkdir output/module_2
mkdir output/module_3
```


Download and create a reference database to BLAST against
```{bash}
## get a text file of ftp paths
Rscript ncbi_refseq_bacteria_ftps.R representative db/ftp_paths.txt

## then download each file
## I used gnu parallel https://www.gnu.org/software/parallel/
## Should take ~20 minutes
parallel --progress --bar -j 8 --gnu wget -q -P db/fna < db/ftp_paths.txt

# Put all of the sequences into 1 file (a few minutes)
gunzip -c db/fna/*.gz | cat > db/fna/representative_seqs.txt

# (optional) delete the gzipped files
rm db/fna/*.gz

## Generate a BLAST database from the downloaded files (~2 minutes)
# Make sure BLAST utilities are downloaded #https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/
makeblastdb \
 -in db/fna/representative_seqs.txt \
 -out db/blast_db/representative_db \
 -dbtype nucl -parse_seqids
```



### Module 1: Subsample the WGS data, then BLAST it to the BLAST database

Input:
 - R1 fastq
Temporary:
 - stats csv
 - R1 subsampled fasta
Output:
 - summary csv
 - read length csv
 - blast csv

```{bash}
## get the basic statistics using seqkit
## https://bioinf.shenwei.me/seqkit/
seqkit stats data/r1.fq -T > output/module_1/basic_stats.txt

## bbmap is used to subsample
#https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbmap-guide/

## subsample 1000 random reads

reformat.sh \
    in="data/r1.fq"  \
    out="data/r1_subsampled.fasta"  \
    samplereadstarget=1000 sampleseed=1234 overwrite=true
    
## Blast The subsample
blastn \
    -task 'megablast' \
    -db db/blast_db/representative_db \
    -max_hsps 3 \
    -max_target_seqs 3000 \
    -perc_identity 95 \
    -query data/r1_subsampled.fasta \
    -outfmt '10 qseqid sseqid length pident mismatch gapopen slen qlen bitscore' \
    -out output/module_1/module_1_blast.csv \
    -num_threads 8
    
## get summary files from the blast results
Rscript module_1.R \
    output/module_1/basic_stats.txt \
    output/module_1/module_1_blast.csv \
    output/module_1 

```

We can view the summary results and check what the best hit was
```{r}
library(data.table)
library(tidyverse)
top_hit <- 
    (fread("output/module_1/module_1_summary_table.csv") %>% 
    pull(accession))[1]

top_hit
```

# get the sequence information from the top hit header
```{bash}
blastdbcmd -db db/blast_db/representative_db -entry NZ_CP053336.1 -outfmt "%t"
```




