---
title: "Reconstruction specs"
output: html_document
---


# Setup

All scripts are expected to be run from the directory where this file is located.

Requirements:
 - BLAST utilities: https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/
 - GNU parallel
 - seqkit: https://bioinf.shenwei.me/seqkit/

## Create directories

Create the base directories
 - `./data` input data folder to reproduce the examples here
 - `./output` output data folder to reproduce the examples here
 - `./db` to store databases required by the scripts

```{bash}
./scripts/setup/create_directories.sh
```

## Download required data

```{bash}
./scripts/setup/download_data.sh
```

## Create reference databases to BLAST against

```{bash}
./scripts/setup/create_blast_databases.sh
```

# Modules

## Module 1: Subsample the WGS data, then BLAST it to the representative BLAST database

Input:
 - R1 fastq
Temporary:
 - stats csv
 - R1 subsampled fasta
Output:
 - summary csv
 - read length csv
 - blast csv

```{bash}

INPUT_FILE=data/r1.fq

# Create basic statistics
seqkit stats $INPUT_FILE -T > output/module_1/basic_stats.txt

# Create a subsample of 1000 random reads
node ./generate-random-reads.js "$INPUT_FILE" "data/r1_subsampled.fasta"

# Blast the subsample
blastn \
    -task 'megablast' \
    -db db/blast_db/representative/db \
    -max_hsps 3 \
    -max_target_seqs 3000 \
    -perc_identity 95 \
    -query data/r1_subsampled.fasta \
    -outfmt '6 qseqid sseqid length pident mismatch gapopen slen qlen bitscore stitle' \
    -out output/module_1/module_1_blast.csv \
    -num_threads 8

# Create summary files from the blast results
Rscript blast_summaries.R \
    output/module_1/basic_stats.txt \
    output/module_1/module_1_blast.csv \
    output/module_1/summary.csv \
    output/module_1/read_length.csv
```

## Module 2: Blast against a specified database

Input:
 - R1 fastq
Temporary:
 - stats csv
 - R1 subsampled fasta
 - titles.txt
 - accessions.txt
 - genus_accessions.txt
Output:
 - summary csv
 - read length csv
 - blast csv

```{bash}

# The top output from Module 1 will indicate which genus should be used as a candidate genome
# We will subset the complete Blast database to extract all sequences corresponding to that sequence

# We choose a genus, which for this example will be salmonella (not case sensitive)
genus="salmonella"

# With the database of a single genus, we blast the R1 reads against them, this time only allowing 100% identity but otherwise the same as Module 1
# Blast The subsample
blastn \
    -task 'megablast' \
    -db "db/blast_db/by_genus/$genus/db" \
    -max_hsps 3 \
    -max_target_seqs 3000 \
    -perc_identity 100 \
    -query data/r1_subsampled.fasta \
    -outfmt '6 qseqid sseqid length pident mismatch gapopen slen qlen bitscore stitle' \
    -out output/module_2/module_2_blast.csv \
    -num_threads 8

# get summary files from the blast results
Rscript blast_summaries.R \
    output/module_1/basic_stats.txt \
    output/module_2/module_2_blast.csv \
    output/module_2/summary.csv \
    output/module_2/read_length.csv
```

### Extract the best hit sequence from the Blast database
```{bash}
# get the top accession from the summary file
acc=$(cut -f2 -d ',' output/module_2/summary.csv | sed -n 2p | sed 's/"//g')
# extract and save the Fasta file
blastdbcmd -db "db/blast_db/by_genus/${genus}/db" -entry "$acc" -outfmt "%f" > \
    data/"$acc".txt
```

## Module 3: Read length optimization

For this step, let's say we give the option to define three read length cutoffs
at 230, 240, and 250.

For the subusample dataset, we'll trim all reads greater than 230 nt to 230 nt,
all greater than 240 nt to 240 nt, and all greater than 250 nt to 250 nt. We won't
look at anything shorter than the lowest cutoff.

```{bash}
# generate read sets at each cutoff
Rscript read_length_optimization.R \
    "230, 240, 250" \
    data/r1_subsampled.fasta \
    output/module_3

# make a directory for cutoff databases
mkdir output/module_3/cutoffs
mkdir output/module_3/blast_results

# create the databases and BLAST
acc=$(cut -f2 -d ',' output/module_2/summary.csv | sed -n 2p | sed 's/"//g')

for CUTOFF in 230 240 250
do
    makeblastdb \
        -in output/module_3/cutoff_$CUTOFF.txt \
        -out output/module_3/cutoffs/cutoff_$CUTOFF -dbtype nucl -parse_seqids
    blastn \
        -task 'megablast' \
        -db output/module_3/cutoffs/cutoff_$CUTOFF \
        -max_hsps 1 \
        -max_target_seqs 5000 \
        -perc_identity 100 \
        -query data/"$acc".txt \
        -outfmt '10 qseqid sseqid length pident mismatch gapopen slen qlen bitscore' \
        -out output/module_3/blast_results/cutoff_$CUTOFF.csv \
        -num_threads 8

done

# summarise the blast results here
Rscript read_length_product.R \
    "230, 240, 250" \
    output/module_3/blast_results

```



